src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -116,79 +123,142 @@ protected override Task UpdateDisplayAsync(in RenderBatch batch)
             //       buffer on every render.
             var batchBytes = MessagePackSerializer.Serialize(batch, RenderBatchFormatterResolver.Instance);
 
-            if (!_client.Connected)
-            {
-                // Buffer the rendered batches while the client is disconnected. We'll send it down once the client reconnects.
-                OfflineRenderBatches.Enqueue(batchBytes);
-                return Task.CompletedTask;
-            }
+            var renderId = Interlocked.Increment(ref _nextRenderId);
+
+            var pendingRender = new PendingRender(
+                renderId,
+                batchBytes,
+                new TaskCompletionSource<object>());
+
+            // Buffer the rendered batches no matter what. We'll send it down immediately when the client
+            // is connected or right after the client reconnects.
 
-            Log.BeginUpdateDisplayAsync(_logger, _client.ConnectionId);
-            return WriteBatchBytes(batchBytes);
+            PendingRenderBatches.Enqueue(pendingRender);
+
+            // Fire and forget the initial send for this batch (if connected). Otherwise it will be sent
+            // as soon as the client reconnects.
+            var _ = WriteBatchBytesAsync(pendingRender);
+
+            return pendingRender.CompletionSource.Task;
         }
 
-        public async Task ProcessBufferedRenderBatches()
+        public Task ProcessBufferedRenderBatches()
         {
-            // The server may discover that the client disconnected while we're attempting to write empty rendered batches.
-            // Discontinue writing in this event.
-            while (_client.Connected && OfflineRenderBatches.TryDequeue(out var renderBatch))
-            {
-                await WriteBatchBytes(renderBatch);
-            }
+            // Upon reconnection we send all the pending batches (we'll let the client sort the order out) 
```

Comment:
Comment should be updated. The client requires the batches to be sent in order, which is what this code does based on the fact that `PendingRenderBatches` entries are always in order.

Reply:
done